File: HW3/problem2/src/FrontEnd.java
import java.util.*;
import java.time.LocalDateTime;

public class FrontEnd {
    private UserInterface ui;
    private BackEnd backend;
    private User user;

    public FrontEnd(UserInterface ui, BackEnd backend) {
        this.ui = ui;
        this.backend = backend;
    }
    
    public boolean auth(String authInfo){
        String[] info = authInfo.split("\n"); // TODO check it's alright
        if(backend.auth(info[0], info[1])){
            user = new User(info[0], info[1]);
            return true;
        }
        return false;
    }

    public void post(Pair<String, String> titleContentPair) {
        Post post = new Post(titleContentPair.key, titleContentPair.value);
        post.setId(backend.getNextPostID(user));
        backend.makePost(user, post);
    }
    
    public void recommend(int N){
        // TODO sub-problem 3
    }

    public void search(String command) {
        // TODO sub-problem 4
    }
    
    User getUser(){
        return user;
    }
}

File: HW6/problem2/frontend.cpp
//
// Created by Jaeyun Kim on 2021/12/18.
//

#include "frontend.h"
#include <sstream>
#include <set>

FrontEnd::FrontEnd(UserInterface & ui_, BackEnd & backend_):ui(ui_), backend(backend_) {
    user = nullptr;
}

FrontEnd::~FrontEnd(){
    if(user) delete user;
}

User* FrontEnd::get_user() {
    return user;
}

bool FrontEnd::auth(const std::string & authInfo) {
    std::string id, pw;
    std::stringstream ss(authInfo);
    ss >> id >> pw;

    user = backend.auth(id, pw);

    return user != nullptr;
}

void FrontEnd::post(const std::pair<std::string, std::string> & titleContentPair) {
    Post post(titleContentPair.first, titleContentPair.second);
    backend.post(user, post);
}

void FrontEnd::recommend(int N) {
    std::vector<Post> post_vec = backend.recommend(user, N);
    for(Post& post : post_vec){
        std::stringstream ss;
        ss << post;
        ui.println(ss.str());
    }
}

void FrontEnd::search(std::vector<std::string> keywords) {
    std::set<std::string> keyword_set;
    for(int i=1; i<keywords.size(); i++){
        keyword_set.insert(keywords[i]);
    }
    std::vector<Post> res = backend.search(keyword_set);

    ui.println("-----------------------------------");
    for(Post & post : res) ui.println(post.get_summary());
}



